### DISCUSSION TITLE
How to stop ImageMagick from eating all available ram and getting OOM'd?

### DISCUSSION THREAD

**Discussion by omentic** [ImageMagick/ImageMagick#7378] [Help]
### ImageMagick version

7.1.1.33

### Operating system, version and so on

Arch, Nix

### Description

I've been attempting to compress a reasonably large (~40 MiB) black-and-white PDF file into something a little smaller. However, on both my laptop and my slightly more powerful desktop (16 GB ram / each), ImageMagick runs for a little while, spikes in memory usage (2GB -> 12GB, about), and gets killed by the OOM killer. 

How can I set a hard memory limit on ImageMagick so this does not happen? `-limit memory`, `MAGICK_AREA_LIMIT`, and `MAGICK_MEMORY_LIMIT` all seem ineffective. The command I've been running is `magick -compress jpeg -quality 75 -units PixelsPerInch -density 300x300 input.pdf output.pdf`.

Any help would be appreciated!

**Comment by fmw42** [comment]
I am not an expert on configuration. But you might check out https://imagemagick.org/script/architecture.php#tera-pixel

**Comment by snibgo** [comment]
The PDF filesize in bytes is not relevant.

Limits on IM's use of memory are not also limits on delegates such as Ghostscript.

How large are the iput.pdf pages, in inches? How many pages? Multiplying by your required density, 300 dpi, gives the total number of pixels. For example: if the pages are 8x11 inches, each page is 8m pixels which shouldn't be a problem, but if you have 1000 pages that would be 8b pixels.

You could convert a few pages at a time, eg `input.pdf[0-9]`.

**Reply by omentic** [reply]
How can I limit the memory use of delegates?
